### PR TITLE
Document module types

### PR DESCRIPTION
**Pull request by marcphilipp**
This PR captures my understanding of the different `ModuleTypes`.

### CODE REVIEW

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt [hunk 1]
```diff
@@ -46,12 +46,56 @@ import java.util.jar.Attributes
...
+    /**
+     * This module type is used internally by default but should not be
+     * referenced by any module.
+     */
     UNDEFINED(JavaVersion.VERSION_1_1),
+
+    /**
+     * This module type is used by modules that contain code that is used by one
+     * of the entry points of using Gradle, such as the Wrapper, the Launcher
+     * and the Tooling API.
```
Comment (pledbrook) [inline buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: You might want to use `<p>` tags for the paragraphs if that's your intention.

Comment (pledbrook) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: Or maybe the behaviour is different with Kotlin source. And of course, if this isn't part of the public API, how the text appears in the online docs is probably irrelevant.

Comment (marcphilipp) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: Yeah, we don't generate docs for this as it is only used internally. Thus, I deliberately left out `<p>` tags. 🙂

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt [hunk 2]
```diff
@@ -215,16 +253,16 @@ class UnitTestAndCompilePlugin : Plugin<Project> {
 open class UnitTestAndCompileExtension(val project: Project) {
     val generatedResourcesDir = project.file("${project.buildDir}/generated-resources/main")
     val generatedTestResourcesDir = project.file("${project.buildDir}/generated-resources/test")
-    var moduleType: ModuleType = ModuleType.UNDEFINED
+    var moduleType: ModuleType? = null
```
Comment (wolfs) [inline buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: Does `lateinit var moduleType: ModuleType` work?

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt [hunk 3]
```diff
@@ -46,12 +46,50 @@ import java.util.jar.Attributes
...
+     * This module type is used by all modules that end up in the core of
+     * the distribution and are not used by entry points or workers.
+     */
     CORE(JavaVersion.VERSION_1_8),
+
+    /**
+     * This module type is used by all modules that become plugins in the
+     * distribution.
+     */
     PLUGIN(JavaVersion.VERSION_1_8),
```
Comment (wolfs) [inline buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: I would merge `PLUGIN` and `CORE` into `CORE`.

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt (hunk 2 above)
Comment (marcphilipp) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: > 'lateinit' modifier is not allowed on properties with a custom getter or setter

Comment (wolfs) [reply buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: 😢

**Pull request merged**